src/main/scala/scorex/core/network/Handshake.scala

```diff
@@ -19,8 +19,75 @@ case class Handshake(applicationName: String,
   require(Option(protocolVersion).isDefined)
 }
 
+
+class NewHandshakeSerializer(featureSerializers: PeerFeature.Serializers,
+                             maxHandshakeSize: Int) extends scorex.core.newserialization.ScorexSerializer[Handshake] {
+
+  override def parse(r: ScorexReader): Handshake = {
+
+    require(r.remaining <= maxHandshakeSize)
+
+    val appName = r.getShortString()
+    require(appName.size > 0)
+
+    val protocolVersion = ApplicationVersionSerializer.parseBytes(
+      r.getBytes(ApplicationVersionSerializer.SerializedVersionLength)
+    ).get
+
+    val nodeName = r.getShortString()
+
+    val fas = r.getInt()
+    val declaredAddressOpt = if (fas > 0) {
+      val fa = r.getBytes(fas - 4)
+      val port = r.getInt()
+      Some(new InetSocketAddress(InetAddress.getByAddress(fa), port))
+    } else None
+
+    val featuresCount = r.getByte()
+    val feats = (1 to featuresCount).flatMap { _ =>
+      val featId = r.getByte()
+      val featBytesCount = r.getShort()
+      val featBytes = r.getBytes(featBytesCount)
+      //we ignore a feature found in the handshake if we do not know how to parse it or failed to do that
+      val featOpt = featureSerializers.get(featId).flatMap { featureSerializer =>
+        featureSerializer.parseBytes(featBytes).toOption
+      }
+      featOpt
+    }
+
+    val time = r.getLong()
+    Handshake(appName, protocolVersion, nodeName, declaredAddressOpt, feats, time)
+  }
+
+  override def serialize(obj: Handshake, w: ScorexWriter): Unit = {
+
+    w.putShortString(obj.applicationName)
+    w.putBytes(obj.protocolVersion.bytes)
+    w.putShortString(obj.nodeName)
+
+    obj.declaredAddress match {
+      case Some(isa) =>
+        val addr = isa.getAddress.getAddress
+        w.putInt(addr.size + 4)
+        w.putBytes(addr)
+        w.putInt(isa.getPort)
+      case None =>
+        w.putInt(0)
+    }
+
+    w.put(obj.features.size.toByte)
```

Comment:
@VictorMikheev Consider a range check here. The ones we have in http://github.com/ScorexFoundation/sigmastate-interpreter/blob/new-coster-rewrite-reducetocrypto/src/main/scala/sigmastate/utils/Extensions.scala#L69 are super handy.

Reply:
fixed